pandas/_libs/tslibs/offsets.pyx

```diff
@@ -777,6 +773,7 @@ cpdef datetime shift_month(datetime stamp, int months, object day_opt=None):
     return stamp.replace(year=year, month=month, day=day)
 
 
+# TODO: Can we declare this so it will take datetime _or_ pandas_datetimestruct
```

Comment:
no please don't functions should not take different types like this. have 2 functions. we specifically generate templates when we have to cover lots of dtypes (not suggesting this at all here though).

Reply:
OK, will remove this comment and address similar things you've mentioned.